File: DATA_STRUCTURES_CPP/Array_(Dynamic_array).cpp
/*Динамический массив - имитация вектора*/
#include <iostream>
#include <vector>

using std::vector;
using std::cout;


template <typename Type>
class Darray {											// Класс описывающий динамический массив
	size_t capac;										// Вместимость
	size_t count;										// Размер массива
	Type *darr;											// Динамический массив
public:
	Darray();											// Конструктор без параметров
	Darray(const size_t &cnt, const Type &elem);		// Конструктор с кол-вом элементов и значением
	Darray(const Darray<Type> &obj);					// Конструктор копирования
	~Darray();											// Деструктор
	void insert(const Type& elm, const size_t& id);		// Вставляет элемент по заданной позиции со сдвигом остальных вправо
	void resize(const size_t &sz, const Type &elm);		// Определяет новый размер массива.
	void reserve(const size_t &sz);						// Резервирует минимальную длину хранилища для объекта
	void push_back(const Type &elm);					// Добавляет элемент в конец
	void erase(const Type &elm);						// Удаляет элемент по значению (без перевыделения памяти, т.е размер массива остается)
	void pop_back();									// Удаляет элемент в конце
	void clear();										// Очищает массив
	void show();										// Печать массива
	bool empty();										// Проверяет, пуст ли массив
	const Type& back() const;							// Возвращает ссылку на последний элемент
	const Type& front() const;							// Возвращает ссылку на первый элемент
	const Type* data() const;							// Возвращает указатель на первый элемент
	const Type& at(const size_t &pos);					// Возвращает ссылку на элемент в заданном положении в массиве.
	size_t capacity() {return capac;}   				// Число элементов, которое массив может содержать без перевыделения
	size_t size()	  {return count;}					// Количество элементов в массиве.
	Type& operator[] (const size_t& id) const;  		// [id]	 - Обращение по индексу
};


template <typename Type>
Darray<Type>::Darray() {
	capac = 2;				// Вместимость - 2 будет по умолчанию
	count = 0;				// Счетчик при создании == 0
	darr  = new Type[2];    // Выделение памяти для 2 элементов типа Type
}

template <typename Type>
Darray<Type>::Darray(const size_t& cnt, const Type& elem) {
	this->Darray<Type>::Darray();
	for (size_t i = 0; i < cnt; i++)
		push_back(elem);
}

template <typename Type>
Darray<Type>::Darray(const Darray<Type>& obj) {
	capac = obj.capac;				// Скопировать вместимость
	count = obj.count;				// Скопировать кол-во элементов
	darr = new Type[count+capac];	// Выделить память для нового массива
	for (size_t i = 0; i < obj.count; ++i)
		darr[i] = obj.darr[i];		// Скопировать элементы
}

template <typename Type>
Darray<Type>::~Darray() {
	clear();
}

template <typename Type>
void Darray<Type>::resize(const size_t& sz, const Type& elm) {
	if (count == 0) {							// Если элементов нет, тогда
		for (size_t i = 0; i < sz; i++)			// Заполнить элементами массив
			push_back(elm);
		return;
	}
	if (sz <= count)							// Если надо перевыделить меньше, чем есть в массиве
		return;									// Ничего не делать, вернуться
	for (size_t i = count; i < sz; i++) {		// Если все ок, то начать за последним элементом в массиве
		push_back(elm);							// И добавить необходимое количество элементов
	}
}

template <typename Type>
void Darray<Type>::reserve(const size_t& sz) {
	if (sz <= capac)								// Если текущая вместитльность больше или равна переданной 
		return;										// Ничего не делать, вернуться, а иначе
	Type *new_arr = new Type[count + capac + sz];	// Создать новый массив, размером старый + новая вместительность
	size_t n = count;								// Запомнить кол_во элементов в массиве
	for (size_t i = 0; i < count; i++)				// Скопировать старый массив в новый
		new_arr[i] = darr[i];
	clear();										// Очистить старый
	darr = new_arr;									// Перекинуть указатель
	capac = sz;										// Вместительность стала равна той, которую передали
	count = n;										// Кол-во элементов осталось столько же
}

template <typename Type>
void Darray<Type>::push_back(const Type& elm) {
	if (capac > 0) {							// Если еще есть элемент в массиве, тогда
		darr[count] = elm;						// Добавить за последним существующим
		capac--;								// Уменьшить вместимость
		count++;								// Увеличить счетчик
	}
	else {										// Если capacity == 0, нужно увеличить массив
		Type *new_array = new Type[count * 2];	// Создать массив в 2 раза больше 
		for (size_t i = 0; i < count; i++)		// Скопировать все элементы в новый массив
			new_array[i] = darr[i];
		delete[] darr;							// Удалить старый массив
		darr = new_array;						// В старый массив положить новый
		darr[count] = elm;						// Добавить элемент в конец
		capac = count - 1;						// Уменьшить вмещаемость
		count++;								// Увеличить счетчик
	}
}

template <typename Type>
void Darray<Type>::insert(const Type& elm, const size_t& id) {
	if (id >= count) {								// Если индекс больше кол-ва элементов, 
		push_back(elm);								// просто вставить элемент в конец массива
		return;										// вернуться
	}
	if (id < 0)										// Если индекс меньше 0 - вернуться
		return;
	Type *tmp_arr = new Type[capac + count + 1];	// Временный массив, для копирования элементов
	bool fl = true;									// Флаг, указывающий на найденный индекс
	for (size_t i = 0; i < count; i++) {			// Обход всего массива
		if (!fl)									// Если false, значит уже добавился элемент по нужному индексу
			tmp_arr[i + 1] = darr[i];				// И тогда нужно делать сдвиг всех элементов на 1 элемент вправо
		if (fl && i != id) {						// Если не дошел до нужного индекса
			tmp_arr[i] = darr[i];					// Копировать элементы в новый массив
		}
		if (fl && i == id) {						// Если индекс совпал, по которому надо вставить элемент
			tmp_arr[i] = elm;						// Добавить его по нужному индексу
			tmp_arr[i + 1] = darr[i];				// Следующий элемент добавляется уэе со сдвигом вправо
			fl = false;								// Поднять флаг, что элемент встретился и выполнялся только 1ый шаг цикла
		}
	}
	size_t n = count, c = capac;					// Запомнить кол-во элементов в массиве + 1 добавленный и вместимость
	this->clear();									// Очистить старый массив
	if (c == 0) {									// Проверить, если он был полон, тогда перевыделить память
		for (size_t i = 0; i < n; i++)				// Пройти в цикле опять, добавить все элементы + перевыделить память
			push_back(tmp_arr[i]);					// Добавление в конец, сохранение порядка
		delete[] tmp_arr;							// Удалить временный массив
		return;										// Вернуться
	}
	darr = tmp_arr;									// Если еще было место в массиве, просто запомнить указатель
	count = n + 1;									// Обновить счетчик, т.к. clear() стер все данные
	capac = c - 1;									// Обновить вместимость
}

template <typename Type>
void Darray<Type>::pop_back() {
	if (count < 1)									// Если массив пуст - вернуться
		return;
	darr[count] = static_cast<Type>(NULL);			// Заполнять удаленные ячейки NULL-ами, и не позволять к ним обращаться потом
	count--;										// Уменьшить счетчик
	capac++;										// Увеличить вмещаемость
}

template <typename Type>
void Darray<Type>::erase(const Type& elm) {
	if (count == 0)
		return;
	size_t id = -1;									// id для найденного элемента
	for (size_t i = 0; i < count; i++) {			// Перебрать массив и найти элемент
		if (elm == darr[i]) {						// Если элемент есть, запомнить id
			id = i;
			break;
		}
	}
	if (id >= 0 && id < count) {					// Если Id корректный, тогда
		Type *tmp_arr = new Type[capac + count];	// Создать временный массив
		for (size_t i = 0; i < count; i++)			// Скопировать в него все значения старого
			tmp_arr[i] = darr[i];					
		tmp_arr[id] = static_cast<Type>(NULL);		// В новом массиве удалить по индексу значение
		size_t n = count;							// Запомнить кол-во элементов
		clear();									// Очистить старый полностью
		for (size_t i = 0; i < n; i++) {			// И заново заполнить уже валидными значениями
			if (tmp_arr[i] != static_cast<Type>(NULL))
				push_back(tmp_arr[i]);
		}
		delete[] tmp_arr;
	}
}

template <typename Type>
void Darray<Type>::clear(){
	while (count != 0)
		pop_back();
}

template <typename Type>
bool Darray<Type>::empty() { return count == 0; }

template <typename Type>
void Darray<Type>::show() {
	if (count < 1)
		return;
	for (size_t i = 0; i < count; i++)
		cout << "[" << i << "] = " << darr[i] << " ";
	cout << '\n';
}

template <typename Type>
const Type& Darray<Type>::back() const {
	return darr[count - 1];
}

template <typename Type>
const Type& Darray<Type>::front() const {
	return darr[0];
}

template <typename Type>
const Type* Darray<Type>::data() const {
	return &darr[0];
}

template <typename Type>
const Type& Darray<Type>::at(const size_t& pos) {
	return darr[pos];
}

template<typename Type>
Type& Darray<Type>::operator [] (const size_t& id) const {
	return darr[id];
}


int main() {
	setlocale(0, "");
	Darray<int> inarr;
	Darray<char> charr;
	Darray<double> flarr;

	cout << "Size : Capacity -> " << inarr.size() << ":" << inarr.capacity() << '\n';
	for (int i = 0; i < 10; i++) {
		cout << inarr.size() << ":";
		cout << inarr.capacity() << "; ";
		inarr.push_back(i * 2);
		charr.insert(char(i + 65), i);
		flarr.push_back(i * 0.24);
	}
	cout << "\n=======================\n";
	cout << "Integer: "; inarr.show();
	cout << "Char:    "; charr.show();
	cout << "Double:  "; flarr.show();
	cout << "Size arr:     " << inarr.size()     << '\n';
	cout << "Capacity arr: " << inarr.capacity() << '\n';
}

File: DATA_STRUCTURES_PY/Array_(Dynamic_array).py
# /*Динамический массив - имитация вектора*/
class Darray:                                   # Класс описывающий динамический массив
    def __init__(self, size, data): pass        # Конструктор
    def push_back(self, elm): pass              # Добавление элемента в конец
    def insert(self, elm, ind): pass            # Вставка элемента по заданному индексу со смещением вправо
    def resize(self, count, elem): pass         # Определяет новый размер массива (заполнение массива элементами)
    def reserve(self, sz): pass                 # Перевыделение вместительности
    def pop_back(self): pass                    # Удаление с конца
    def erase(self, elem): pass                 # Удаление по значению
    def clear(self): pass                       # Очистка массива
    def empty(self): pass                       # True если пуст, иначе false
    def back(self): pass                        # Возвращает последний элемент массива
    def front(self): pass                       # Возвращает первый элемент массива
    def data(self): pass                        # Возвращает адрес 1-го элемента
    def at(self, ind): pass                     # Возвращает значение массива по индексу
    def __getitem__(self, ind): pass            # Получить значение по индексу через [id]
    def __setitem__(self, ind, value): pass     # Заменить значение по индексу [id] = value
    def show(self): pass                        # Печать массива
    def size(self): pass                        # Размер массива
    def capacity(self): pass                    # Вместительность массива
"""-------------------------------------------------------------------------"""

class Darray:
    def __init__(self, size=2, data=None):
        if data is None:                        # Если при создании данных нет, тогда
            self._size = 0                      # Счетчик элементов == 0, т.к элементов нет
            self._capacity = size               # А вместительность равна либо 2 либо, переданной в параметре
        if data is not None:                    # Если надо заполнить массив значениями, тогда
            self._size = size                   # Счетчик элементов актуален и равен либо дефолту, либо заданному
            self._capacity = 0                  # Равна 0, потом будет перевыделение памяти при необходимости
        self._array = [data] * size             # Заполнить массив значениями

    def push_back(self, elm):                   # Добавление элемента в конец
        if self._capacity > 0:                  # Если еще есть элемент в массиве, тогда
            self._array[self._size] = elm		# Добавить за последним существующим
            self._size += 1         			# Увеличить счетчик
            self._capacity -= 1				    # Уменьшить вместимость
        else:                                   # Если capacity == 0, нужно увеличить массив,
            # В старый массив добавить еще пространства увеличенного вдове (по типу вектора)
            self._array = self._array + ([None] * (self._size * 2))
            self._array[self._size] = elm       # Добавить элемент
            self._capacity = self._size - 1     # Уменьшить вместительность
            self._size += 1                     # Увеличить счетчик элементов

    def insert(self, elm, ind):                 # Вставка элемента по заданному индексу со смещением вправо
        if ind >= self._size:                   # Если индекс >= кол-ву элементов в массиве, тогда
            self.push_back(elm)                 # Вставить в конец
            return                              # И вернуться
        if ind < 0:                             # Если индекс кривой
            return                              # Ничего не делать
        new_arr = self._array + [None]          # Создать временный массив, для смещения всех элементов после индекса
        flag = True                             # Флаг, указывающий на найденный индекс после которого нужно смещение
        for i in range(self._size):             # Поэлементный обход всего массива
            if flag is False:                   # Если флаг опущен, значит надо делать смещение вправо на 1
                new_arr[i + 1] = self._array[i] # Сдвиг на 1 элемент вправо
            if flag and i != ind:               # Если флаг поднят и индекс ещё не найден, тогда
                new_arr[i] = self._array[i]     # Делать обычное копирование
            if flag and i == ind:               # Если встретился индекс, по которому нужно добавить элемент
                new_arr[i] = elm                # Добавить этот элемент
                new_arr[i + 1] = self._array[i] # Добавить элемент по i индексу уже со сдвигом вправо
                flag = False                    # Опустить флаг, чтобы дальше добавлять со смещением

        if self._capacity == 0:                 # Проверить, если старый массив был заполнен, тогда
            self._array = new_arr               # Скопировать новый массив со сдвигом на 1 вправо
            del new_arr                         # Удалить временный массив, он не нужен
            self._capacity = 0                  # Вместительность также 0, т.е. он по-прежнему полон
            self._size += 1                     # Увеличился счетчик на 1 добавленный элемент
            return                              # Вернуться
        # Если еще было место в массиве, тогда
        self._array = new_arr                   # Скопировать новый массив со сдвигом на 1 вправо
        del new_arr                             # Удалить его
        self._capacity -= 1                     # Уменьшить вместительность
        self._size += 1                         # Увеличить счетчик

    def resize(self, count, elem):              # Определяет новый размер массива (заполнение массива элементами)
        if count <= self._size or count < 0:    # Если кол-во перевыделяемой вместительности < 0 или размера массива
            return                              # Ничего не делать
        if self._size == 0:                     # Если элементов в массиве нет, тогда
            for i in range(count):              # Заполнить его элементами
                self.push_back(elem)
            return
        for i in range(self._size, count):      # Если все корректно, тогда начать за последним элементом
            self.push_back(elem)                # Добавлять необходимое кол-во элементов

    def reserve(self, sz):                      # Перевыделение вместительности
        if sz <= self._capacity:                # Если текущая вместительность больше или равна заданной
            return                              # Ничего не делать
        self._array = self._array + [None] * sz # Увеличить вместительность текущего массива
        self._capacity = sz                     # Вместительность стала равной заданной

    def pop_back(self):                         # Удаление с конца
        if self._size == 0:                     # Если элементов нет
            return
        self._array[self._size-1] = None       # Заполнить ячейку null и не позволять обращаться к ним
        self._size -= 1
        self._capacity += 1

    def erase(self, elem):                      # Удаление по значению
        if self._size == 0:                     # Если нет элементов, вернуться
            return
        ind = -1                                # Индекс найденного элемента
        for i in range(self._size):             # Пройти по массиву
            if elem == self._array[i]:          # Запомнить индекс элемента
                ind = i
                break
        if (ind >= 0) and (ind < self._size):   # Если индекс корректный, элемент был найден
            self._array[ind] = None             # Стереть элемент
            tmp_arr = []                        # Временный массив для сдвига элементов
            for i in range(self._size):         # Пройти па массиву
                if self._array[i] is not None:  # Скопировать поэлементно
                    tmp_arr.append(self._array[i])
            self._size -= 1                     # Уменьшить размер
            self._capacity += 1                 # Вместительность увеличить
            self._array = tmp_arr               # Новый массив без удаленного элемента
            del tmp_arr                         # Удалить ненужный

    def clear(self):
        while self._size != 0:
            self.pop_back()

    def empty(self):
        return self._size == 0

    def back(self):
        return self._array[self._size - 1] if self._size != 0 else None

    def front(self):
        return self._array[0] if self._size != 0 else None

    def data(self):
        return id(self._array[0]) if self._size != 0 else None

    def at(self, ind):
        return self._array[ind] if self._size > ind else None

    def __getitem__(self, ind):
        return self._array[ind] if self._size > ind else None

    def __setitem__(self, ind, value):
        if ind >= self._size:
            return
        self._array[ind] = value

    def show(self):
        if self._size == 0:
            print("Array is empty!")
            return
        print('[ ', end='')
        for i in range(self._size):
            print(self._array[i], end=', ')
        print(']')

    def size(self):
        return self._size

    def capacity(self):
        return self._capacity


def main():
    inarr = Darray(5, 3)
    charr = Darray(5, 'w')
    flarr = Darray()

    for i in range(15):
        inarr.push_back(i)
        charr.insert(chr(i+67), i)
        flarr.push_back(round(i * 0.28, 2))

    inarr.show()
    charr.show()
    flarr.show()

    for i in range(5):
        inarr.erase(i)
        charr.pop_back()
    inarr.show()
    charr.show()


if __name__ == "__main__":
    main()